tests/src/3d/testqgs3dutils.cpp

```diff
@@ -157,5 +159,39 @@ void TestQgs3DUtils::testQgsBox3DDistanceTo()
   }
 }
 
+void TestQgs3DUtils::testQgsRay3D()
+{
+  QgsRay3D ray( QVector3D( 0, 0, 0 ), QVector3D( 1, 1, 1 ) );
+  float t = 1.0f + ( float )( rand() % 1000 ) / 1000.0f;
+  QVector3D p1 = ray.origin() + t * ray.direction();
+  QVector3D p2 = ray.origin() - t * ray.direction();
+  // point already on the ray
+  QVERIFY( ray.projectedPoint( p1 ) == p1 );
+  QVERIFY( ray.projectedPoint( p2 ) == p2 );
+
+  // t >= 0 then the point is in front of the ray
+  QVERIFY( ray.isInFront( p1 ) );
+  // t < 0 then the point is in front of the ray
+  QVERIFY( !ray.isInFront( p2 ) );
+
+  for ( int i = 0; i < 8; ++i )
+  {
+    // random vector
+    QVector3D n = QVector3D( 1.0f + ( float )( rand() % 1000 ) / 1000.0f, 1.0f + ( float )( rand() % 1000 ) / 1000.0f, 1.0f + ( float )( rand() % 1000 ) / 1000.0f ).normalized();
+    // random point on the ray
+    float t = 1.0f + ( float )( rand() % 1000 ) / 1000.0f;
```

Comment:
If possible at all, I would prefer that we do not create random numbers and do any extra math... Best tests are when we have some fixed numbers (they can be randomly chosen, but not during the runtime :-) ) and then fixed numbers which are the expected results... With no for-cycles, no extra math, basically to take any complexity away from tests, to lower the chance of bugs in the tests :-) ... and if we have let's say 2-3 different values that should be fine, maybe add some edge cases if you can think of some.

Also, with floating point numbers, I believe Qt uses fuzzy comparison in QCOMPARE, so it should be ok to just do QCOMPARE( a, b )

This would be an example of a better test:
```
QCOMPARE( ray.projectedPoint( QVector3D( 321, 654, 987 ) ), QVector3D( 123, 456, 789 ) );
```

Reply:
I changed to the approach you suggested, I got a bit carried away with maths :-)
and somwhow QCOMPARE doesn't do fuzzy comarison properly for QVector3D so I ended up comparing components individually.